ui-v2/app/components/discovery-chain.js

```diff
@@ -0,0 +1,235 @@
+        // subset this just shifts the subset of the front of the URL for the moment
+        const temp = item.ID.split('.');
+        temp.shift();
+        resolver.ID = temp.join('.');
+        resolver.Subsets.push(item);
+      }
+      if (typeof node.Resolver.Failover !== 'undefined') {
+        // FIXME: Figure out if we can get rid of this
+        /* eslint ember/no-side-effects: "warn" */
+        set(failoverable, 'Failover', targetsToFailover(node.Resolver.Failover.Targets, item.ID));
+      }
+    }
+  });
+  return Object.values(resolvers);
+};
+const getFailovers = function(nodes = {}) {
+  const failovers = [];
+  Object.values(nodes)
+    .filter(item => item.Type === 'resolver')
+    .forEach(function(item) {
+      (get(item, 'Resolver.Failover.Targets') || []).forEach(failover => {
+        failovers.push(failover);
+      });
+    });
+  return failovers;
+};
+const getType = function(nodes = {}, type) {
+  return Object.values(nodes).filter(item => item.Type === type);
+};
+
+const targetsToFailover = function(targets, a) {
+  let type;
+  const Targets = targets.map(function(b) {
+    // FIXME: this isn't going to work past namespace for services
+    // with dots in the name
+    const [aRev, bRev] = [a, b].map(item => item.split('.').reverse());
+    const types = ['Datacenter', 'Namespace', 'Service', 'Subset'];
+    return bRev.find(function(item, i) {
+      const res = item !== aRev[i];
+      if (res) {
+        type = types[i];
+      }
+      return res;
+    });
+  });
+  return {
+    Type: type,
+    Targets: Targets,
+  };
+};
+
+export default Component.extend({
+  dom: service('dom'),
+  dataStructs: service('data-structs'),
+  classNames: ['discovery-chain'],
+  classNameBindings: ['active'],
+  selectedId: '',
+  x: 0,
+  y: 0,
+  tooltip: '',
+  activeTooltip: false,
+  splitters: computed('chain.Nodes', function() {
+    return getType(get(this, 'chain.Nodes'), 'splitter').map(function(item) {
+      set(item, 'ID', `splitter:${item.Name}`);
+      return item;
+    });
+  }),
+  routers: computed('chain.Nodes', function() {
+    // Right now there should only ever be one 'Router'.
+    return getType(get(this, 'chain.Nodes'), 'router');
+  }),
+  routes: computed('routers', function() {
+    return get(this, 'routers').reduce(function(prev, item) {
+      return prev.concat(
+        item.Routes.map(function(route, i) {
+          return {
+            ...route,
+            ID: `route:${item.Name}-${i}`,
+          };
+        })
+      );
+    }, []);
+  }),
+  nodeResolvers: computed('chain.Nodes', function() {
+    return getNodeResolvers(get(this, 'chain.Nodes'));
+  }),
+  resolvers: computed('nodeResolvers.[]', 'chain.Targets', function() {
+    return getTargetResolvers(get(this, 'chain.Targets'), this.nodeResolvers);
+  }),
+  graph: computed('chain.Nodes', function() {
+    const graph = this.dataStructs.graph();
+    Object.entries(get(this, 'chain.Nodes')).forEach(function([key, item]) {
```

Comment:
I wouldn't have thought so, I've no idea if ember polyfills it either. I know ember ember polyfills a bunch of stuff as we've spoken about it before.

I've never managed to find any ember documentation that gives me a list of what it polyfills and is therefore safe to use, I don't suppose you know of any do you?

I think right now we are kind of sweeping IE11 under the rug 😄 We have quite a few IE11 💥 's

I would generally use `keys` for this, but since we have `entries` now (that is polyfillable) I'm kinda trying it out instead of using `keys`.

Reply:
As far as I know, there is no documentation like that. They treat those extension methods as part of the API, so you can see it on the `Array` and `String` doc pages, but it doesn't explicitly call out what is a prototype extension.

But Ember definitely doesn't extend the `Object` prototype, "only" `String`, `Array`, and `Function`. And the function extensions are just the convenience methods `on`, `property`, and `observes`.

Reply:
Ah k, thanks for the extra info!